### PR TITLE
Offer to add await if converting from Task<T> to T

### LINKED ISSUE #5137: Offer "Add Await" codefix for CS0029
Instead of showing a “Quick Action” on line 14 (int myInt = MyIntMethodAsync()), it shows an error.  But all the other lines 10 and 12, it properly shows that `await` should be added.
```C#
using System.Threading.Tasks;

namespace WpfApplication1
{
       class TestClass
       {
              private async Task MyTestMethod1Async()
              {
                      await MyObjectMethodAsync();
                      MyObjectMethodAsync();

                      MyIntMethodAsync();

                      int myInt = MyIntMethodAsync();
              }

              private Task<object> MyObjectMethodAsync()
              {
                      return Task.FromResult(result: (object)null);
              }

              private Task<int> MyIntMethodAsync()
              {
                      return Task.FromResult(result: 1);
              }

       }
}
```

### PR DESCRIPTION
**Pull request by jmarolf**
If we see CS0029, AND the initial type is Task<T> AND the expected type is T, and the method is async: offer to add await to the expression.
Fixes #5137

### CODE REVIEW

**Comment by jmarolf** [comment]
Reviewers: @balajikris, @brettfo, @CyrusNajmabadi, @davkean, @dpoeschl, @jasonmalinowski, @	DustinCampbell, @Pilchie, @rchande

**Comment by dpoeschl** [comment]
Is there a bug for this?

**Comment by jmarolf** [comment]
retest this please

File: src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs [hunk 1]
```diff
@@ -66,17 +79,81 @@ protected override Task<SyntaxNode> GetNewRoot(SyntaxNode root, SyntaxNode oldNo
...
+            }
+
             INamedTypeSymbol taskType = null;
             INamedTypeSymbol returnType = null;
-            return TryGetTypes(expression, semanticModel, out taskType, out returnType) &&
+            return TryGetTaskAndExpressionTypes(expression, semanticModel, out taskType, out returnType) &&
             semanticModel.Compilation.ClassifyConversion(taskType, returnType).Exists;
         }
 
+        private static bool DoesExpressionReturnGenricTaskWhoseArgumentsMatchLeftSide(ExpressionSyntax expression, SemanticModel semanticModel, Project project, CancellationToken cancellationToken)
```
Comment (bkoelman) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: typo?
DoesExpressionReturnGenricTaskWhoseArgumentsMatchLeftSide
DoesExpressionReturnGenericTaskWhoseArgumentsMatchLeftSide

File: src/Features/VisualBasic/Portable/CodeFixes/Async/VisualBasicAddAwaitCodeFixProvider.vb [hunk 2]
```diff
@@ -53,10 +57,65 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeFixes.Async
             End Select
         End Function
 
-        Private Function IsCorrectReturnType(expression As ExpressionSyntax, semanticModel As SemanticModel) As Boolean
+        Private Function DoesExpressionReturnGenricTaskWhoseArgumentsMatchLeftSide(expression As ExpressionSyntax, semanticModel As SemanticModel, project As Project, cancellationToken As CancellationToken) As Boolean
+            If expression Is Nothing Then
```
Comment (bkoelman) [inline src/Features/VisualBasic/Portable/CodeFixes/Async/VisualBasicAddAwaitCodeFixProvider.vb]: same typo as above in DoesExpressionReturnGenricTaskWhoseArgumentsMatchLeftSide

File: src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs [hunk 3]
```diff
@@ -66,17 +79,81 @@ protected override Task<SyntaxNode> GetNewRoot(SyntaxNode root, SyntaxNode oldNo
...
+        private static bool IsInAsyncFunction(ExpressionSyntax expression)
+        {
+            foreach (var node in expression.Ancestors())
+            {
+                switch (node.Kind())
+                {
+                    case SyntaxKind.ParenthesizedLambdaExpression:
+                    case SyntaxKind.SimpleLambdaExpression:
+                    case SyntaxKind.AnonymousMethodExpression:
+                        return (node as AnonymousFunctionExpressionSyntax)?.AsyncKeyword.Kind() == SyntaxKind.AsyncKeyword;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: i'd prefer   !...IsMissing

File: src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs (hunk 1 above)
Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: Thanks for catching this!

File: src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs (hunk 3 above)
Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: good idea.

**Comment by jmarolf** [comment]
retest mac

**Comment by jmarolf** [comment]
test mac please

**Comment by jmarolf** [comment]
retest mac please

**Comment by jmarolf** [comment]
test mac

**Comment by jmarolf** [comment]
test mac please

File: src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs [hunk 4]
```diff
@@ -66,17 +79,81 @@ protected override Task<SyntaxNode> GetNewRoot(SyntaxNode root, SyntaxNode oldNo
...
+        private static bool IsInAsyncFunction(ExpressionSyntax expression)
+        {
+            foreach (var node in expression.Ancestors())
+            {
+                switch (node.Kind())
+                {
+                    case SyntaxKind.ParenthesizedLambdaExpression:
+                    case SyntaxKind.SimpleLambdaExpression:
+                    case SyntaxKind.AnonymousMethodExpression:
+                        return (node as AnonymousFunctionExpressionSyntax)?.AsyncKeyword.IsMissing == true;
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: What's with `IsMissing == true`?  Shouldn't it be false?

**Comment by Pilchie** [comment]
:+1:

File: src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs (hunk 4 above)
Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/Async/CSharpAddAwaitCodeFixProvider.cs]: Converting from bool? to bool

**Pull request merged**